tools/generator/templates/player/caml/interface.cc.jinja2

```diff
@@ -0,0 +1,238 @@
+}
+template <>
+value cxx_to_caml<value, bool>(bool in)
+{
+    CAMLparam0();
+    CAMLreturn(Val_int(in));
+}
+template <typename Cxx>
+value cxx_to_caml_array(const std::vector<Cxx>& in)
+{
+    CAMLparam0();
+    CAMLlocal1(v);
+    size_t size = in.size();
+    if (size == 0)
+        CAMLreturn(Atom(0));
+    v = caml_alloc(size, 0);
+    for (size_t i = 0; i < size; ++i)
+        caml_initialize(&Field(v, i), cxx_to_caml<value, Cxx>(in[i]));
+    CAMLreturn(v);
+}
+template <typename CamlType, typename Cxx>
+Cxx caml_to_cxx(CamlType in)
+{
+    return in.__if_that_triggers_an_error_there_is_a_problem;
+}
+template<>
+std::string caml_to_cxx<value, std::string>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(std::string, String_val(in));
+}
+template <>
+int caml_to_cxx<value, int>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(int, Int_val(in));
+}
+template <>
+double caml_to_cxx<value, double>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(double, Double_val(in));
+}
+template <>
+bool caml_to_cxx<value, bool>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(bool, Int_val(in));
+}
+template <typename Cxx>
+std::vector<Cxx> caml_to_cxx_array(value in)
+{
+    CAMLparam1(in);
+    std::vector<Cxx> out;
+    mlsize_t size = Wosize_val(in);
+    for (size_t i = 0; i < size; ++i)
+        out.push_back(caml_to_cxx<value, Cxx>(Field(in, i)));
+    CAMLreturnT(std::vector<Cxx>, out);
+}
+{% for enum in game.enum %}
+
+{{ enum.enum_summary|cxx_comment }}
+template<>
+value cxx_to_caml<value, {{ enum.enum_name }}>({{ enum.enum_name }} in)
+{
+    CAMLparam0();
+    CAMLreturn(Val_int(in));
+}
+
+template<>
+{{ enum.enum_name }} caml_to_cxx<value, {{ enum.enum_name }}>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT({{ enum.enum_name }}, ({{ enum.enum_name }})Int_val(in));
+}
+{% endfor %}
+{% for struct in game.struct %}
+
+{{ struct.str_summary|cxx_comment }}
+template<>
+value cxx_to_caml<value, {{ struct.str_name }}>({{ struct.str_name }} in)
+{
+    CAMLparam0();
+    CAMLlocal1(out);
+    out = caml_alloc({{ struct.str_field|length }}, 0);
+    {% for field_name, field_type, field_comment in struct.str_field %}
+    caml_initialize(&Field(out, {{ loop.index0 }}), {{ field_type|cxx_to_caml}}(in.{{ field_name }}));
+    {% endfor %}
+    CAMLreturn(out);
+}
+
+template<>
+{{ struct.str_name }} caml_to_cxx<value, {{ struct.str_name }}>(value in)
+{
+    CAMLparam1(in);
+    {{ struct.str_name }} out;
+    {% for field_name, field_type, field_comment in struct.str_field %}
+    out.{{ field_name }} = {{ field_type|caml_to_cxx }}(Field(in, {{ loop.index0 }}));
+    {% endfor %}
+    CAMLreturnT({{ struct.str_name }}, out);
+}
+{% endfor %}
+
+/*
+** Inititialize caml
+*/
+static inline void _init_caml()
+{
+    static bool is_initialized = false;
+    if (!is_initialized)
+    {
+        is_initialized = true;
+        const char* argv[2] = {"./caml", NULL};
+        caml_startup(const_cast<char**>(argv));
```

Comment:
This is not great. Consider using:

```
char program_name[] = "./caml";
char* argv[2] = {program_name, NULL};
caml_startup(argv);
```

Yes, it's annoying, but it's the proper way to declare what you want (allocate the string on the stack then deallocate it after the startup, which can change its arguments).

Reply:
Old generator again, and I agree, I frowned when I copy/pasted it.